lib/segment/src/index/field_index/map_index/mutable_map_index.rs

```diff
@@ -0,0 +1,159 @@
+        if self.point_to_values.len() <= idx as usize {
+            return Ok(());
+        }
+
+        let removed_values = std::mem::take(&mut self.point_to_values[idx as usize]);
+
+        if !removed_values.is_empty() {
+            self.indexed_points -= 1;
+        }
+        self.values_count -= removed_values.len();
+
+        for value in &removed_values {
+            if let Some(vals) = self.map.get_mut(value) {
+                vals.remove(&idx);
+            }
+            let key = MapIndex::encode_db_record(value, idx);
+            self.db_wrapper.remove(key)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn get_db_wrapper(&self) -> &DatabaseColumnWrapper {
+        &self.db_wrapper
+    }
+
+    pub fn load_from_db(&mut self) -> OperationResult<bool> {
+        if !self.db_wrapper.has_column_family()? {
+            return Ok(false);
+        }
+        self.indexed_points = 0;
+        for (record, _) in self.db_wrapper.lock_db().iter()? {
+            let record = std::str::from_utf8(&record).map_err(|_| {
+                OperationError::service_error("Index load error: UTF8 error while DB parsing")
+            })?;
+            let (value, idx) = MapIndex::decode_db_record(record)?;
+            if self.point_to_values.len() <= idx as usize {
+                self.point_to_values.resize(idx as usize + 1, Vec::new())
+            }
+            if self.point_to_values[idx as usize].is_empty() {
+                self.indexed_points += 1;
+            }
+            self.values_count += 1;
+
+            let entry = self.map.entry(value);
+            self.point_to_values[idx as usize].push(entry.key().clone());
+            entry.or_default().insert(idx);
+        }
+        Ok(true)
+    }
+
+    pub fn get_values(&self, idx: PointOffsetType) -> Option<&[N]> {
+        self.point_to_values.get(idx as usize).map(|v| v.as_slice())
+    }
+
+    pub fn get_indexed_points(&self) -> usize {
+        self.indexed_points
+    }
+
+    pub fn get_values_count(&self) -> usize {
+        self.values_count
+    }
+
+    pub fn get_unique_values_count(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn get_points_with_value_count<Q>(&self, value: &Q) -> Option<usize>
+    where
+        Q: ?Sized,
+        N: std::borrow::Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.map.get(value).map(|p| p.len())
+    }
+
+    pub fn get_iterator<Q>(&self, value: &Q) -> Box<dyn Iterator<Item = PointOffsetType> + '_>
+    where
+        Q: ?Sized,
+        N: std::borrow::Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.map
+            .get(value)
+            .map(|ids| Box::new(ids.iter().copied()) as Box<dyn Iterator<Item = PointOffsetType>>)
+            .unwrap_or_else(|| Box::new(iter::empty::<PointOffsetType>()))
+    }
+
+    pub fn get_values_iterator(&self) -> Box<dyn Iterator<Item = &N> + '_> {
+        Box::new(self.map.keys())
+    }
+}
```

Comment:
Maybe it is `get_keys_iterator`?

Reply:
scratch that, `get_values_iterator` is correct

Reply:
Good catch but `values` because Value here is related to the map index value, not a `self.map` container